FaceIterator.h

```diff
@@ -0,0 +1,424 @@
+#ifdef USE_MPI
+        MPI_Comm_rank(m_puml.comm(), &rank);
+        MPI_Comm_size(m_puml.comm(), &commSize);
+
+        const auto& faces = puml.faces();
+
+        std::vector<std::vector<int>> transfer(commSize);
+        for (int i = 0; i < faces.size(); ++i) {
+            int lid[2];
+            Upward::cells(puml, faces[i], lid);
+            assert(!(lid[0] == -1 && lid[1] != -1));
+            assert(faces[i].shared().size() <= 1);
+            if (lid[0] != -1 && faces[i].isShared()) {
+                transfer[faces[i].shared()[0]].push_back(i);
+            }
+        }
+
+        m_transferSize = std::vector<int>(commSize);
+        m_transferDisp = std::vector<int>(commSize+1);
+        m_transferDisp[0] = 0;
+        for (int i = 0; i < commSize; ++i) {
+            m_transferSize[i] = transfer[i].size();
+            m_transferDisp[i+1] = m_transferDisp[i]+m_transferSize[i];
+            std::sort(transfer[i].begin(), transfer[i].end(),
+                [&faces] (int a, int b) -> bool {
+                    return faces[a].gid() < faces[b].gid();
+                });
+        }
+
+        m_transferCell = std::vector<int>(m_transferDisp[commSize]);
+        m_transferFace = std::vector<int>(m_transferDisp[commSize]);
+        for (int i = 0, j = 0; i < commSize; ++i) {
+            for (int k = 0; k < m_transferSize[i]; ++j, ++k) {
+                const auto& face = faces[transfer[i][k]];
+                int lid[2];
+                Upward::cells(puml, face, lid);
+                m_transferCell[j] = lid[0];
+                m_transferFace[j] = transfer[i][k];
+            }
+        }
+#endif
+    }
+    
+    // FaceHandlerFunc: void(int,int,const T&,const T&)
+    template<typename T, typename FaceHandlerFunc,
+        std::enable_if_t<std::is_invocable_v<FaceHandlerFunc, int, int, const T&, const T&>, bool> = true>
+    void forEach(const std::vector<T>& cellData,
+                    FaceHandlerFunc&& faceHandler,
+                    MPI_Datatype mpit = MPITypeInfer<T>::type()) {
+        auto cellHandler = [&cellData](int fid, int cid){return cellData[cid];};
+        forEach<T, T>(cellHandler, faceHandler, [](int a, int b){}, mpit);
+    }
+
+    // FaceHandlerFunc: void(int,int,const T&,const T&)
+    // BoundaryFaceHandlerFunc: void(int,int)
+    template<typename T, typename FaceHandlerFunc, typename BoundaryFaceHandlerFunc,
+        std::enable_if_t<std::is_invocable_v<FaceHandlerFunc, int, int, const T&, const T&>, bool> = true,
+        std::enable_if_t<std::is_invocable_v<BoundaryFaceHandlerFunc, int, int>, bool> = true>
+    void forEach(const std::vector<T>& cellData,
+                    FaceHandlerFunc&& faceHandler,
+                    BoundaryFaceHandlerFunc&& boundaryFaceHandler,
+                    MPI_Datatype mpit = MPITypeInfer<T>::type()) {
+        auto cellHandler = [&cellData](int fid, int cid){return cellData[cid];};
+        forEach<T, T>(cellHandler, faceHandler, boundaryFaceHandler, mpit);
```

Comment:
One more thing: Would be good to use std::forward here for faceHandler and boundaryFaceHandler. (Likewise for the other forEach functions.)

Reply:
Thanks again for the suggestion. It has been included in e60c0027272812ca6c9b7e66c31abe6ad4e8549e .